diff --git a/src/alloc-override.c b/src/alloc-override.c
index b5109de..aef5fba 100644
--- a/src/alloc-override.c
+++ b/src/alloc-override.c
@@ -258,7 +258,7 @@ extern "C" {
   void*  valloc(size_t size)               { return mi_valloc(size); }
   void   vfree(void* p)                    { mi_free(p); }
   size_t malloc_good_size(size_t size)     { return mi_malloc_good_size(size); }
-  int    posix_memalign(void** p, size_t alignment, size_t size) { return mi_posix_memalign(p, alignment, size); }
+  //int    posix_memalign(void** p, size_t alignment, size_t size) { return mi_posix_memalign(p, alignment, size); }
 
   // `aligned_alloc` is only available when __USE_ISOC11 is defined.
   // Note: it seems __USE_ISOC11 is not defined in musl (and perhaps other libc's) so we only check
