diff --git a/include/mimalloc/internal.h b/include/mimalloc/internal.h
index eae85ab..0c3f07e 100644
--- a/include/mimalloc/internal.h
+++ b/include/mimalloc/internal.h
@@ -34,7 +34,11 @@ terms of the MIT license. A copy of the license can be found in the file
 #define mi_decl_hidden
 #elif (defined(__GNUC__) && (__GNUC__ >= 3)) || defined(__clang__) // includes clang and icc
 #define mi_decl_noinline        __attribute__((noinline))
+#if defined(__MVS__)
+#define mi_decl_thread
+#elif
 #define mi_decl_thread          __thread
+#endif
 #define mi_decl_cache_align     __attribute__((aligned(MI_CACHE_LINE)))
 #define mi_decl_weak            __attribute__((weak))
 #define mi_decl_hidden          __attribute__((visibility("hidden")))
